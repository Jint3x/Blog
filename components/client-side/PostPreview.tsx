import React from "react"


interface PostPreview_I {
    thumbnail: string,
    title: string,
    description: string,
    date: string, 
    tags: string[],
    url: string,

    // If loading is set to true, all other fields won't be used
    loading?: boolean,
}

function PostPreview(props: PostPreview_I) {
    return (
        <div className={"w-72 bg-white ml-auto mr-auto h-80 rounded shadow-lg shadow shadow-md shadow-personal mt-12 mb-12 overflow-hidden"}>

            {
                props.loading ? 
                (
                    <div className={"h-20per bg-black"} />
                )
                :
                (
                    <div 
                     style={{backgroundImage: `url(${props.thumbnail})`}} 
                     className={"h-20per bg-center overflow-hidden cursor-pointer"} 
                     onClick={() => location.assign(props.url)}
                    />
                )
            }
            

            <div className={"h-65per color-default-dark relative"}>
                <h3 className={"text-center pt-1 pb-2 text-2xl overflow-hidden"}>
                    {
                        props.loading ? "Loading..." : props.title
                    }
                </h3>

                <div className={"relative h-3/6 pl-2 pr-2 mt-2"}>
                    <p className={"overflow-hidden h-full"}>
                        {props.loading ? "" : props.description}
                    </p>
                    <div className={"absolute bottom-0 h-12 w-full bg-gradient-to-t from-white to-transparent"} />
                </div>
                

                <p className={"absolute bottom-1 text-sm left-1 text-post-date"}>
                    {props.loading ? "" : props.date}
                </p>
            </div>

            <div 
             className={"h-15per bg-main-blue-lighter flex-nowrap flex items-center overflow-hidden relative cursor-pointer"}
             onClick={() => location.assign(props.url)}
            >
                <div className={"absolute h-full w-1/12 right-0 bg-gradient-to-l from-main-blue-lighter to-transparent"} />
                {
                    props.loading ? "" : props.tags.map(tag => <RenderTag name={tag} key={tag} />)
                }
            </div>
        </div>
    )
}


interface RenderTag_I {
    name: string,
}

function RenderTag(props: RenderTag_I) {
    return (
        <p className={"bg-main-green-trans ml-2 mr-2 text-sm rounded-md p-1 no-break whitespace-nowrap"}>
            {props.name}
        </p>
    )
}




export default PostPreview