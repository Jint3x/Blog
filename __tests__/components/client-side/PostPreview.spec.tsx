/**
 * @jest-environment jsdom
 */
 import PostPreview from "../../../components/client-side/PostPreview";
 import ReactRenderer from "react-test-renderer"
 import { screen, render } from "@testing-library/react"


test("Renders all the text successfully", () => {
    render(
        <PostPreview 
         thumbnail={"https://example.com/asd"}
         title={"A title"}
         description={"A test description"}
         date={"2nd of Feb, 2021"}
         tags={["Deep Dive", "Technology"]}
         url={"/blog/how-to-test"}
        />
    )

    let title = screen.getByText("A title");
    let description = screen.getByText("A test description");
    let date = screen.getByText("2nd of Feb, 2021");
    let tagOne = screen.getByText("Deep Dive");
    let tagTwo = screen.getByText("Technology");

    expect(title.nodeName).toBe("H3");
    expect(description.nodeName).toBe("P")
    expect(date.nodeName).toBe("P")
    expect(tagOne.nodeName).toBe("P")
    expect(tagTwo.nodeName).toBe("P")
})


test.todo("Make sure the links are correctly navigating when clicked.")


test("Snaphsot of the PostPreview component", () => {
    let tree = ReactRenderer.create(
        <PostPreview 
            thumbnail={"https://example.com/asd"}
            title={"A title"}
            description={"A test description"}
            date={"2nd of Feb, 2021"}
            tags={["Deep Dive", "Technology"]}
            url={"/blog/how-to-test"}
        />
    )
    .toJSON();

    expect(tree).toMatchSnapshot();
})