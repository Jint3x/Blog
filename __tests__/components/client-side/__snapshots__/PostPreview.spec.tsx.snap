// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snaphsot of the PostPreview component 1`] = `
<div
  className="w-72 ml-auto mr-auto h-80 rounded shadow-lg shadow shadow-md shadow-personal mt-12 mb-12 overflow-hidden"
>
  <div
    className="h-20per bg-center overflow-hidden cursor-pointer"
    onClick={[Function]}
    style={
      Object {
        "backgroundImage": "url(https://example.com/asd)",
      }
    }
  />
  <div
    className="h-65per color-default-dark relative"
  >
    <h3
      className="text-center pt-1 pb-2 text-2xl overflow-hidden"
    >
      A title
    </h3>
    <div
      className="relative h-3/6 pl-2 pr-2 mt-2"
    >
      <p
        className="overflow-hidden h-full"
      >
        A test description
      </p>
      <div
        className="absolute bottom-0 h-12 w-full bg-gradient-to-t from-white to-transparent"
      />
    </div>
    <p
      className="absolute bottom-1 text-sm left-1 text-post-date"
    >
      2nd of Feb, 2021
    </p>
  </div>
  <div
    className="h-15per bg-main-blue-lighter flex-nowrap flex items-center overflow-hidden relative cursor-pointer"
    onClick={[Function]}
  >
    <div
      className="absolute h-full w-1/12 right-0 bg-gradient-to-l from-main-blue-lighter to-transparent"
    />
    <p
      className="bg-main-green-trans ml-2 mr-2 text-sm rounded-md p-1 no-break whitespace-nowrap"
    >
      Deep Dive
    </p>
    <p
      className="bg-main-green-trans ml-2 mr-2 text-sm rounded-md p-1 no-break whitespace-nowrap"
    >
      Technology
    </p>
  </div>
</div>
`;
